File: python_guidelines/algorithms/recursion/binary-search/binary-search.py
"""
When the sequence is sorted and indexable we can use
binary search that runs in O(log n) time, whereas sequential
search runs in O(n) time.
"""

def binary_search(data, target, low, high):
  """
  Return True if target is found in indicated portion of a Python list.
  """
  
  if low > high:
    return False
  else:
    mid = (low + high) / 2
    if target == data[mid]:
      return True
    elif target < data[mid]:
      # recur on the portion left of the middle
      return binary_search(data, target, low, mid - 1)
    else:
      # recur on the portion right of the middle
      return binary_search(data, target, mid + 1, high)

File: container/AppAcademy/Curriculum/12 Career Quest/W23D1 Binary Search and Linked Lists/binary_search.js
// Given a sorted array of numbers and a target num, return a boolean indicating whether or not that target is contained in the array.

// Programmatically, we want to satisfy the following behavior:

// binarySearch([5, 10, 12, 15, 20, 30, 70], 12);  // => true
// binarySearch([5, 10, 12, 15, 20, 30, 70], 24);  // => false

function binarySearch (array, target) {
  //if target not found, return false
  if (array.length === 0) return false

  let midIdx = Math.floor(array.length / 2)
  let leftSide = array.slice(0, midIdx)
  let rightSide = array.slice(midIdx + 1)

  if (target < array[midIdx]) {
    return binarySearch(leftSide, target)
  } else if (target > array[midIdx]) {
    return binarySearch(rightSide, target)
  } else {
    // if target is neither less than or great than, then must be target
    return true
  }
}

console.log(binarySearch([5, 10, 12, 15, 20, 30, 70], 12))  // => true
console.log(binarySearch([5, 10, 12, 15, 20, 30, 70], 24))  // => false

// Time Complexity: O(log(n))
// n is the length of the input array
// We have no loops, so we must only consider the number of recursive calls it takes to hit the base case
// The number of recursive calls is the number of times we must halve the array until it's length becomes 0. This number can be described by log(n)
// for example, say we had an array of 8 elements, n = 8
// the length would halve as 8 -> 4 -> 2 -> 1
// it takes 3 calls, log(8) = 3
// Space Complexity: O(n)
// Our implementation uses n space due to half arrays we create using slice. Note that JavaScript slice creates a new array, so it requires additional memory to be allocated.

// When should we use Binary Search?
// Use this algorithm when the input data is sorted!!! This is a heavy requirement, but if you have it,you'll have an insanely fast algorithm.